tensorflow/core/grappler/optimizers/data/hoist_data_discarding_ops.cc

```diff
@@ -0,0 +1,120 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/core/grappler/optimizers/data/hoist_data_discarding_ops.h"
+
+#include "absl/container/flat_hash_set.h"
+#include "tensorflow/core/framework/attr_value.pb.h"
+#include "tensorflow/core/framework/node_def.pb.h"
+#include "tensorflow/core/grappler/clusters/cluster.h"
+#include "tensorflow/core/grappler/grappler_item.h"
+#include "tensorflow/core/grappler/mutable_graph_view.h"
+#include "tensorflow/core/grappler/op_types.h"
+#include "tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.h"
+#include "tensorflow/core/grappler/optimizers/data/function_utils.h"
+#include "tensorflow/core/grappler/optimizers/data/graph_utils.h"
+#include "tensorflow/core/grappler/utils.h"
+#include "tensorflow/core/platform/protobuf.h"
+
+namespace tensorflow {
+namespace grappler {
+namespace {
+
+constexpr std::array<const char*, 3> kDataDiscarding = {
+    "ShardDataset", "SkipDataset", "TakeDataset",
+};
+
+constexpr std::array<const char*, 6> kCardinalityPreserving = {
```

Comment:
There are other cardinality preserving transformations in the core API and it would be good if your implementation works for those too: `concatenate`, `enumerate`, `shuffle`, and `zip`.

Reply:
As for `concatenate`, I think it's hard to hoist those ops when we don't know the cardinality figure. And for `shuffle`, I think that the hoisting may change the ouput. I can add support for `zip` (`enumerate`) I guess, however, in that case, it will be a condition `IsZip` and some extra implementation because right now the supported ops are all unary. I wonder if we could first make the basic functionality work and then let me create another pr for zip?